### PR TITLE
SOLR-49: Handle SOLR errors in response writers

### PR DESCRIPTION
**Pull request by samj1912**
Still figuring out the correct way to deal with this.

Currently this returns a blank response when an invalid search field is given. Do we have a response format for errors in Search? If yes, I would like to extract the error and simply display that when it is present.

Not sure if my way is the most optimal since currently it just hides the problem.

### CODE REVIEW

**Comment by mayhem** [comment]
I guess the real questions to ask here are:

1. Can we catch the error and find out which field is not supported? (seems yes)
2. What mechanisms to we have for reporting errors?

If we can do both, then that would be the ideal situation.

**Comment by samj1912** [comment]
I am not quite certain what the correct way to display errors is, I did try throwing a runtime exception with a proper message but it appears as a server error (500) as opposed to an API response error (which is what is shown for other formats like xml or json).

**Comment by mayhem** [comment]
Let's see if mineo has input. If not, ask the solr community how to handle this.

File: src/main/java/org/musicbrainz/search/solrwriter/MBXMLWriter.java [hunk 1]
```diff
@@ -314,13 +314,31 @@ private static void adjustScore(float maxScore, Object object,
 		}
 	}
 
+    private void writeResponse(Writer writer, MetadataListWrapper metadatalistwrapper)
```
Comment (mineo) [inline src/main/java/org/musicbrainz/search/solrwriter/MBXMLWriter.java]: Minor nitpick: These lines are indented by 4 spaces, the rest uses tabs.

File: src/main/java/org/musicbrainz/search/solrwriter/MBXMLWriter.java [hunk 2]
```diff
@@ -314,13 +314,31 @@ private static void adjustScore(float maxScore, Object object,
...
+
 	public void write(Writer writer, SolrQueryRequest req, SolrQueryResponse res)
 			throws IOException {
 		MetadataListWrapper metadatalistwrapper = new MetadataListWrapper();
 
 		NamedList vals = res.getValues();
 
 		ResultContext con = (ResultContext) vals.get("response");
+		if (con == null)
+		{
```
Comment (mineo) [inline src/main/java/org/musicbrainz/search/solrwriter/MBXMLWriter.java]: The old search server has count and offset both set to 0 in this case (http://search.musicbrainz.org/ws/2/area/?query=Area:Berlin), so you might need an additional call to `metadatalistwrapper.setCountAndOffset(0, 0);` here.

**Comment by samj1912** [comment]
Fixed the response now to match with the current search server.

**Comment by samj1912** [comment]
This is what happens when I create a Solr exception for this 

image
 
any thoughts on how I should wrap this into an xml object?

I dont have a predefined class in mmd schema for errors.

**Comment by samj1912** [comment]
Fixed it after a lot of hair pulling (java may or may not have been responsible for this)

image
image

**Comment by samj1912** [comment]
You can test it here - http://35.193.119.46:8983/solr/artist/select?q=asasas:*

http://35.193.119.46:8983/solr/artist/select?q=asasas:*&wt=mbjson

File: src/main/java/org/musicbrainz/search/solrwriter/MBXMLWriter.java [hunk 3]
```diff
@@ -321,11 +333,35 @@ private static void adjustScore(float maxScore, Object object,
...
+		StringWriter sw = new StringWriter();
+
+		Exception solrError = res.getException();
+
+		if (solrError != null)
+		{
+			MBError mbError = new MBError();
+			int code = Integer.parseInt(res.getResponseHeader().get("status").toString());
+
+			mbError.setCode(code);
```
Comment (mineo) [inline src/main/java/org/musicbrainz/search/solrwriter/MBXMLWriter.java]: I'm not sure if this is OK with JAXB, but maybe add a constructor to MBError that takes 2 arguments and sets code and message directly?

File: src/main/java/org/musicbrainz/search/solrwriter/MBXMLWriter.java [hunk 4]
```diff
@@ -321,11 +333,35 @@ private static void adjustScore(float maxScore, Object object,
...
+			writer.write(sw.toString());
+			return;
+		}
+		
 		MetadataListWrapper metadatalistwrapper = new MetadataListWrapper();
 
 		NamedList vals = res.getValues();
 
 		ResultContext con = (ResultContext) vals.get("response");
+
```
Comment (mineo) [inline src/main/java/org/musicbrainz/search/solrwriter/MBXMLWriter.java]: I don't think this empty line is necessary, but that's no blocker :)

File: src/main/java/org/musicbrainz/search/solrwriter/MBError.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package org.musicbrainz.search.solrwriter;
+
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+import javax.xml.bind.annotation.XmlType;
+
+@XmlRootElement(name = "error")
+@XmlType(propOrder = {"code", "message"})
+public class MBError {
```
Comment (mineo) [inline src/main/java/org/musicbrainz/search/solrwriter/MBError.java]: Maybe add a small docstring.

File: src/main/java/org/musicbrainz/search/solrwriter/MBXMLWriter.java (hunk 3 above)
Comment (samj1912) [reply src/main/java/org/musicbrainz/search/solrwriter/MBXMLWriter.java]: I tried that, doesn't seem to work :(

**Pull request merged**